File: C2/PDOO/Practicas/DeepSpaceJava/src/deepspace/Dice.java
package deepspace;



import java.util.Random;
/**
 *
 * @author Pedro Pablo Ruiz Huertas y Juan Antonio Villegas Recio
 */
public class Dice {
    private float NHANGARSPROB;
    private float NSHIELDSPROB;
    private float NWEAPONSPROB;
    private float FIRSTSHOTPROB;
    private float EXTRAEFFICIENCYPROB;
    private Random generator;
    
    Dice(){
        NHANGARSPROB=(float) 0.25;
        NSHIELDSPROB=(float) 0.25;
        NWEAPONSPROB=(float) 0.33;
        FIRSTSHOTPROB=(float) 0.5;
        EXTRAEFFICIENCYPROB=(float) 0.8;
        
        generator=new Random();
        
    }
    
    public boolean extraEfficiency(){
        float prob;
        prob=generator.nextFloat();
        
        return prob<EXTRAEFFICIENCYPROB;     
    }
    
    public int initWithNHangars(){
        float prob;
        prob = generator.nextFloat();
        
        if(prob<NHANGARSPROB){
            return 0;
        }
        else{
            return 1;
        }
    }
    
    public int initWithNWeapons(){
        float prob;
        prob=generator.nextFloat();
        
        if(prob<NWEAPONSPROB){
            return 1;
        }
        else{
                    if(prob<(2*NWEAPONSPROB)){
                        return 2;
                    }
                    else{
                        return 3;
                    }
        }
        
    }
    
    
    public int initWithNShields(){
        float prob;
        prob = generator.nextFloat();
        
        if(prob<NSHIELDSPROB){
            return 0;
        }
        else{
            return 1;
        }
    }
    
    public int whoStarts(int nPlayers){
        int player;
        player=generator.nextInt(nPlayers);
        return player;
    }
    
    GameCharacter firstShot(){
        float prob;
        prob=generator.nextFloat();
        if(prob<FIRSTSHOTPROB){
            return GameCharacter.SPACESTATION;
        }
        else{
            return GameCharacter.ENEMYSTARSHIP;
        }
    }
    
    boolean spaceStationMoves(float speed){
        float prob;
        prob=generator.nextFloat();
        return prob<speed;
        
    }
    
    public String toString(){
        String mensaje = "El numero de hangares inicial es "+this.initWithNHangars()+
                ", el numero de armas inicial es "+this.initWithNWeapons()+
                ", el numero de escudos inicial "+this.initWithNShields()+
                ", y dispara primero "+this.firstShot();
        return mensaje;
    }
}

File: C2/PDOO/Practicas/DeepSpaceRuby/lib/Dice.rb

#@author Pedro Pablo Ruiz Huertas y Juan Antonio Villegas Recio

#encoding:utf-8

require_relative 'GameCharacter.rb'

module Deepspace
  class Dice
    
    public
    
    def initialize
      @NHANGARSPROB = 0.25;
      @NSHIELDSPROB = 0.25;
      @NWEAPONSPROB = 0.33;
      @FIRSTSHOTPROB = 0.5;
      @EXTRAEFFICIENCYPROB = 0.8;
    end
    
    def initWithNHangars
      prob = rand
      if prob < @NHANGARSPROB
        return 0
      else
        return 1
      end
    end
    
     def initWithNWeapons
       prob = rand
       if prob < @NWEAPONSPROB
        return 1
      else
        if prob < 2*@NWEAPONSPROB
          return 2
        else 
          return 3
        end
      end
    end
    
    def initWithNShield
      prob = rand
      if prob < @NSHIELDSPROB
        return 0
      else
        return 1
      end
    end
    
    def extraEfficiency
      prob = rand
      if prob < @EXTRAEFFICIENCYPROB
        return false
      else
        return true
      end
    end
    
    def whoStarts(new_nPlayers)
      player = rand(new_nPlayers)
      return player
    end
    
    def firstShot
      prob = rand
      if prob < @FIRSTSHOTPROB
        return GameCharacter::SPACESTATION
      else
        return GameCharacter::ENEMYSTARSHIP
      end
    end
    
    def spaceStationMoves(new_speed)
      prob = rand
      return prob < new_speed
    end
    
  end
end
